agent/rpc/subscribe/auth.go

```diff
@@ -0,0 +1,40 @@
+package subscribe
+
+import (
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/consul/stream"
+)
+
+// EnforceACL takes an acl.Authorizer and returns the decision for whether the
+// event is allowed to be sent to this client or not.
+func enforceACL(authz acl.Authorizer, e stream.Event) acl.EnforcementDecision {
+	switch {
+	case e.IsEndOfSnapshot(), e.IsEndOfEmptySnapshot():
+		return acl.Allow
+	}
+
+	switch p := e.Payload.(type) {
+	case state.EventPayloadCheckServiceNode:
+		csn := p.Value
+		if csn.Node == nil || csn.Service == nil || csn.Node.Node == "" || csn.Service.Service == "" {
+			return acl.Deny
+		}
+
+		// TODO: what about acl.Default?
+		// TODO(streaming): we need the AuthorizerContext for ent
+		if dec := authz.NodeRead(csn.Node.Node, nil); dec != acl.Allow {
+			return acl.Deny
+		}
+
+		// TODO(streaming): we need the AuthorizerContext for ent
+		// Enterprise support for streaming events - they don't have enough data to
+		// populate it yet.
+		if dec := authz.ServiceRead(csn.Service.Service, nil); dec != acl.Allow {
```

Comment:
```suggestion
		var authzCtx acl.AuthorizerContext
		csn.Service.FillAuthzContext(&authCtx)

		if dec := authz.NodeRead(csn.Node.Node, &authzCtx); dec != acl.Allow {
			return acl.Deny
		}

		// Enterprise support for streaming events - they don't have enough data to
		// populate it yet.
		if dec := authz.ServiceRead(csn.Service.Service, &authzCtx); dec != acl.Allow {
```

I am not entirely sure yet whether the underlying types are the same and this will actually work but this is generally how its done in other areas.

Reply:
Thanks! This code seems to work. I've been wondering if there isn't already some code somewhere that does this authorization for a CheckServiceNode. It might make sense to have them use the same function

Reply:
I don't believe a common check exists anywhere. There are a number of acl helpers in agent/acl.go but none for this use case. I think having methods on the types for validating read/write access would be good for the CheckServiceNode type and many others. Similar to what we have for config entries:

https://github.com/hashicorp/consul/blob/364f6589c89f146b1617842b03c7c4ed2227b157/agent/structs/config_entry.go#L42-L45

None of that exists though.

Reply:
Nice! I can add that.  I notice these do not take an `AuthorzierContext`, even though one would need to be created for `ResolveTokenAndDefaultMeta`. Is the context that is passed to `ResolveTokenAndDefaultMeta` captured by the authorizer, so that it doesn't have to be passed to these `CanRead`/`CanWrite` ?

Reply:
For Config Entries and plenty of other data, all the information to populate the authz context with is stored in the object so the authz context is created internally.

https://github.com/hashicorp/consul/blob/0c9177f6a5ddd5eb9cf58a36ea9f98f56197c8c4/agent/structs/config_entry.go#L136-L140

Reply:
Ahh, I had completed missed the `FillAuthzContext` line. Thanks! This makes sense now.